### PR TITLE
Configure Websocket routes

### PR DESCRIPTION
**Pull request by rshriram**
**What this PR does / why we need it**:
Configures websocket routes in Envoy for routes marked for websocket upgrade

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #
#1006 
**Special notes for your reviewer**:

**Release note**:

```release-note
Configure websocket routes in Envoy
```

### CODE REVIEW

File: proxy/envoy/config.go [hunk 1]
```diff
@@ -503,6 +503,19 @@ func buildInboundListeners(mesh *proxyconfig.ProxyMeshConfig, sidecar proxy.Node
 				Routes:  []*HTTPRoute{route},
 			}
 
+			// Websocket enabled routes need to have an explicit use_websocket : true
+			// This setting needs to be enabled on Envoys at both sender and receiver end
+			if protocol == model.ProtocolHTTP {
+				// get all the route rules applicable to the instances
+				rules := config.RouteRulesByDestination(instances)
```
Comment (kyessenov) [inline proxy/envoy/config.go]: Hm, the target generally doesn't know where the request comes from, so can't truthfully infer the rules that carried the request to destination unless it's in the request metadata.

Comment (rshriram) [inline proxy/envoy/config.go]: That is fine.. All we need is to know the URIs that support websockets and setup special route entries in envoy with websocket enabled. We don't have to do source based routing or any form of matching. This is just a roundabout way to collect websocket enabled URLs. I was hoping that we would put in some caching in that config object such that we dont have to compute this again and again.

File: test/client/client.go [hunk 2]
```diff
@@ -105,12 +108,59 @@ func makeHTTPRequest(client *http.Client) func(int) func() error {
...
+				log.Printf("[%d] Body=%s\n", i, msg)
+			}
+
+			conn, _, err := client.Dial(url, req)
+			if err != nil {
+				// timeout or bad handshake
+				return err
+			}
+			// nolint: errcheck
+			defer conn.Close()
```
Comment (ijsnellf) [inline test/client/client.go]: does `// nolint: errcheck` have to be on the same line?

File: proxy/envoy/config.go [hunk 3]
```diff
@@ -518,19 +518,41 @@ func buildInboundListeners(mesh *proxyconfig.ProxyMeshConfig, sidecar proxy.Node
 		// services' kubeproxy to our specific endpoint IP.
 		switch protocol {
 		case model.ProtocolHTTP, model.ProtocolHTTP2, model.ProtocolGRPC:
-			route := buildDefaultRoute(cluster)
+			defaultroute := buildDefaultRoute(cluster)
```
Comment (ijsnellf) [inline proxy/envoy/config.go]: nit: golang style `defaultroute` -> `defaultRoute`

File: test/client/client.go (hunk 2 above)
Comment (rshriram) [reply test/client/client.go]: err on.. ths worked..

File: test/grpcecho/echo.proto [hunk 4]
```diff
@@ -16,14 +16,14 @@ syntax = "proto3";
 
 package grpecho;
 
-service EchoTestService {
-  rpc Echo(EchoRequest) returns (EchoResponse);
+service GrpcEchoTestService {
```
Comment (ldemailly) [inline test/grpcecho/echo.proto]: @rshriram why is this changed ?
is that what causes
https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_pilot/1102/pilot-presubmit/574/

Comment (ldemailly) [reply test/grpcecho/echo.proto]: filled #1112 as I don't understand how it seems to have passed in this pr and not passing anymore

Comment (ldemailly) [reply test/grpcecho/echo.proto]: @rshriram ping - why the change ?
